tests/testthat/test-missing_data.R

```diff
@@ -5,31 +5,139 @@ test_that("defMiss produces correct output", {
+  #form1 <- paste0(form1val0, "*(1-m) + m*", form1val1)
+  
+  defM <- defMiss(varname = "x1", formula = hardProbForm, logit.link = FALSE)
+  defM <- defMiss(defM, varname = "x2", formula = form1, logit.link = FALSE)
+  defM <- defMiss(defM, varname = "u", formula = 1, logit.link = FALSE) # not observed
+  
+  missMat <- genMiss(dtAct1, defM, idvars = "id")
+  
+  ## check probabilistic column has correct distribution
+  hardProbAct <- mean(missMat[, x1])
+  hPDiff <- abs(hardProbAct - hardProbForm)
+  
+  expect_true(hPDiff < 0.02)
+  
+  ## check dependent column 1
+  #check probs when m is 0/1, u is 0/1
+  
+  mIs0 <- dtAct1[, m == 0]
+  mIs1 <- dtAct1[, m == 1]
+
+  missMat0 <- missMat[mIs0]
+  missMat1 <- missMat[mIs1]
+  
+  avg0 <- mean(missMat0[, x2])
+  avg1 <- mean(missMat1[, x2])
+  
+  diff0 <- abs(form1val0 -avg0)
+  diff1 <- abs(form1val1 -avg1)
+  
+  expect_true(diff0 < 0.02 && diff1 < 0.02)
+  
+  ## check empty column
+  expect_true(all(missMat[, u] == 1))
+  
+  # repeated (not worth looking at if !baseline and !mono)
+  ## includes lags
+  
+  ## baseline
+  #if missing at baseline, missing at all other perios
+  idNum <- 500
+  
+  dtAct2 <- genData(idNum, def1)
+  dtAct2 <- trtObserve(dtAct2, formulas = .5, logit.link = FALSE, grpName = "rx")
+  
+  defLong <- defDataAdd(varname = "y", dist = "normal", formula = "10 + period*2 + 2 * rx", variance = 2)
+  
+  dtTime <- addPeriods(dtAct2, nPeriods = 4)
+  dtTime <- addColumns(defLong, dtTime)
+  
+  defMlong <- defMiss(varname = "x1", formula = .20, baseline = TRUE)
+  defMlong <- defMiss(defMlong,varname = "y", formula = "-1.5 - 1.5 * rx + .25*period", logit.link = TRUE, baseline = FALSE, monotonic = FALSE)
+  
+  missMatLong <- genMiss(dtTime, defMlong, idvars = c("id","rx"), repeated = TRUE, periodvar = "period")
+  
+  casevec <- NULL
+  for(i in 1:idNum) {
+    tempMM <- missMatLong[id == i]
+    if(tempMM[1, x1] == 1) {
+      casevec <- c(casevec, all(tempMM[, x1] == 1))
+    }
+  }
+  
+  expect_true(all(casevec == TRUE))
+  
+  ### monotonic
+  #same as baseline stipulation
+  defMlong <- defMiss(varname = "x1", formula = .20, baseline = TRUE)
+  defMlong <- defMiss(defMlong,varname = "y", formula = "-1.8 - 1.5 * rx + .25*period", logit.link = TRUE, baseline = FALSE, monotonic = TRUE)
+  
+  missMatLong <- genMiss(dtTime, defMlong, idvars = c("id","rx"), repeated = TRUE, periodvar = "period")
+  
+  casevec <- NULL
+  for(i in 1:idNum) {
+    tempMM <- missMatLong[id == i]
+    # TODO is there a cleaner way to do this?
+    if(tempMM[1, x1] == 1) {
+      casevec <- c(casevec, all(tempMM[, x1] == 1))
+    } else if(tempMM[2, x1] == 1) {
+      casevec <- c(casevec, (all(tempMM[-1, x1] == 1) && tempMM[1, x1] == 0))
+    } else if(tempMM[3, x1] == 1) {
+      casevec <- c(casevec, (all(tempMM[-(1:2), x1] == 1) && tempMM[1:2, x1] == 0))
+    } else if(tempMM[4, x1] == 1) {
+      casevec <- c(casevec, (all(tempMM[-(1:3), x1] == 1) && tempMM[1:3, x1] == 0))
+    }
+  }
```

Comment:
I think we can improve on this - as you suspected, but I haven't thought about it yet.